README.md
# [RUNALYZE v4.3.0] Release with minor fixes/changes.

This fork of Runalyze is a release fork with all needed dependencies and can be used directly with [Docker](https://github.com/2er0/runalyze-docker).
I have done some small fixes/imporvements.
Because it based on the [Release 4.3.0](https://github.com/Runalyze/Runalyze/releases/tag/v4.3.0) i hope it is more future-proof in an "old" docker container.

I host it on a private Pine64 Rock64 SOC computer to host my family activities (running, walking, mountain climbing, swimming). It runs on a Debian Buster/ARM64 in a Docker container serviced with docker-compose. Buster supports PHP 7.3 and it runs without problems with some PHP warnings. As input GPS devices i use Garmin Forerunner 45S, Garmin Fenix 6 and Android-Handy with ApeMap/OruxMaps. I import my tacks without use of Garmin-Tools (like Garmin Connect) and so i think no of my private sensible health-data is transmit to the "public" cloud.

With my other Github project [Clone of Tkl2Gpx](https://github.com/codeproducer198/Tkl2Gpx) i have imported my old running activities from the year 2012 until now into RUNALYZE. These old tracks are record with a GPS MapJack watch and transformed to GPX files imported via RUNYLZE bulk-job.

Here some fixes/improvements i have done in RUNALYZE (see details in the commits):
* Fixes some small bugs until the base release is running on my environment (missing DB attribute, wrong/missing number values, ...)
* Fixes some small bugs while importing FIT files from Garmin
* Batch/Bulk-imports can now set/override the sports type
* Imports from MapJack watch/GPX and Garmin FR45 & Fenix6/FIT results in errors because missing heart-rates and altitutes. Now the NULL will be filled.
* Sport types hiking and (new) mountain climbing.
* Imported filename is stored in title attribute.
* Temperature of FIT files are stored in the temp attribute as average value.
* 2020-09-27: Import Garmin FIT "total_training_effect"-attribute (Aerob Training Effect) already with greater 0.0 (and not even 1.0)
* 2020-09-27: Garmin FIT "total_anaerobic_training_effect" attribute as "Anaerobic Training Effect" is imported
	* Store it in the DB (runalyze_training.fit_anaerobic_training_effect)
	* Added on the dataset configuration
	* Also add field on statistic heart-rate view (activity main page)
	* **Migration 20200926230800 is necessary!**
* 2020-09-29: Garmin FIT attributes as fit_lactate_threshold_hr, fit_total_ascent/descent (the watch original values) are imported
	* Store it in the DB (runalyze_training.fit_anaerobic_training_effect)
	* Shown in the statistic view (activity main page) under 'Miscellaneous' panel
	* **Migration 20200928150400 is necessary!**
* 2020-09-29: Show further FIT file attributes (recovery-time, training-effect, creator, vo2max, performance-conditions, hvr) on statistic view (activity main page) under 'Miscellaneous' panel
* 2020-10-06: Loading weather data now uses the time in the middle of the activity _(start + (duration / 2))_.
* 2020-10-07: Add support for [meteostat](https://meteostat.net) historical weather data while editing a activity or while bulk import. Usable with setting the new "meteostatnet_api_key" in config.yml.
* 2020-10-17 to 2020-11-04: Import heartrate and temperature of Fenix 6 for swimming activities.
* 2020-11-04: Some fixing of correlate trackdata to laps/swim-lanes.
* 2020-11-09: Auto detection of type "interval-training" (detection only works in batch/bulk-mode). You must configure a training-type with short-cut "IT" to your sports in the configuration to use this feature.
* 2021-01-02: Fix "Start date" of an existing equipment is set to the stored date (not to current date).
* 2021-01-02: While batch/bulk-importing of activities main equipment of the sports are assigned to new activities
	* only main equipment-type are considered (set your equipment type as main equipment in the sports configuration)
	* only equipment-type with single-choice considered
	* a unique/time-ranged equipment must exists for this sport; if multiple equipments are found, nothing is assigned and a warning is logged while importing
	* keep in mind, that f. e. multiple shoes can not be mapped because no shoe can be clearly identified
* 2021-01-10: Add API-REST service /api/import/activity for uploading and importing activities for an account
	* The POST request must be a "multipart/form-data" to support multiple times in one request.
	* This new API uses the existing command ActivityBulkImportCommand
	* If no problems occur and all uploaded files are imported successfully a HTTP state 200 is returned otherwise (duplicates or failed) a state 202 is returned.
	* If there are technical trouble (while creating the tmp-folder, moving files ...) a HTTP state 500 is returned.
	* Additional infos and the output of the "runalyze:activity:bulk-import" are set in the response content as "Content-Type: text/plain"
	* Example curl to use: curl -k -u "<runalyze-account>:<pwd>" -H "Accept-Language: de,en" -X POST https://<domain>/api/import/activity -F 'file1=@afile.fit' -F 'file2=@another.fit' [-w 'HttpCode: %{http_code}\n']
* 2021-11-05: Add tests(-folder) from the archived repository to ensure dependency changes
	1. database preperation script `tests/scripts/preparetests.sh` (one time action per DB instance)
	2. wrapper script for all tests `tests/scripts/runtests.sh`
	* some tests has errors (i think is caused by the usage of PHP7); see `runtests.sh` for expected failures
	* `tests` folder can be excluded for production environment
* 2021-11-09: Upgrade PHP dependencies to newer/recent versions
	* only deps where none or only few code changes necessary
	* changes affected composer-files, vendor folder, database migrations, tests
	* Clone `laufhannes/GpxTrackPoster` to `codeproducer198/GpxTrackPoster` and use this dependency (commit from Mar 11, 2017)
* 2021-11-14: Upgrade web dependencies to newer/recent versions
	* only javascript deps where no error occur while the resource build or the app works
	* changes affected `bower.json`, `web/vendor` folder
* 2021-11-14: Upgrade to a newer Runalyze Glossary version (fixed commit-id dev-master#d7e2540bf51dc1aabe4dbdce85af96e8203311b6)
	* Seems the last version which works with the v4.3.0
* 2021-11-14: Garmins _Performance Condition_ is now saved in the purpose column for the newest watches
	* _Performance Condition_ in the "fit detail" section of an selected activity is now also displayed as a value between -20/+20
	* Supported watches are: Forerunner 640 & 645 & 935 & 945 & 735, Fenix 3 & 5 & 6
	* Caution: the search page still uses values based on 100; if you search for "+5" use "105" ;-)
	* The value is stored so far in the column `runalyze_training.fit_hrv_analysis`; to fix your database you can use this sqls to move the values to `fit_performance_condition`:
	`update runalyze_training set fit_performance_condition = fit_hrv_analysis where fit_performance_condition is null and fit_hrv_analysis is not null and fit_performance_condition_end is not null;`
	and
	`update runalyze_training set fit_hrv_analysis = null where fit_performance_condition = fit_hrv_analysis is not null and fit_performance_condition_end is not null;`

Please notice:
* All the changes are only done for me to use this great product for me.
* I don't take any responsibility if you running this version on your infrastructure and have problems.
* The extensions i made was done on a release version. So i do not build a release. I have no translations for the new features.

## Database migration

For Migration of the Database use the commands:
- Check state: `/usr/bin/php /app/runalyze/bin/console doctrine:migrations:status --env=prod --show-versions`
- Do it: `/usr/bin/php /app/runalyze/bin/console doctrine:migrations:migrate --env=prod [--dry-run]`
- Rollback to an previous version: `/usr/bin/php /app/runalyze/bin/console doctrine:migrations:migrate --env=prod <VersionAufDieZurückgerolltWerdenSoll>`
Notice: Do the migration with your "project" user.


CHANGELOG.md
## Changelog

A complete list of our releases with all minor bugfix versions (vX.Y.*) can be found
at [github](https://github.com/Runalyze/Runalyze/releases)
and in our [blog](https://blog.runalyze.com/releases/).

* for recent changes see the `README.md`
* v4.3, [update instructions](https://docs.runalyze.com/en/latest/upgrade/4.x.html), 28.01.2018
* [v4.2](https://blog.runalyze.com/releases/4-2-0/), [update instructions](https://docs.runalyze.com/en/latest/upgrade/4.x.html#upgrade-from-4-1-to-4-2), 12.08.2017: minor release (trend analysis, climb score, glossary) 
* [v4.1](https://blog.runalyze.com/releases/4-1-0/), [update instructions](https://docs.runalyze.com/en/latest/upgrade/4.x.html#upgrade-from-4-0-to-4-1), 07.05.2017: minor release (refactored sports configuration, new navigation, retina maps)
* [v4.0](https://blog.runalyze.com/releases/4-0-0/), [update instructions](https://docs.runalyze.com/en/latest/upgrade/4.x.html#upgrade-from-3-2-to-4-0), 12.03.2017: minor release (poster tool, Removal of VDOT features)
* [v3.2](https://blog.runalyze.com/releases/3-2-0/), [update instructions](https://docs.runalyze.com/en/latest/upgrade/3.x.html#upgrade-from-3-1-to-3-2), 29.01.2017: minor release (ANOVA tool, geohashline, refactoring)
* [v3.1](https://blog.runalyze.com/allgemein-en/minor-changes-in-runalyze-v3-1/), [update instructions](https://docs.runalyze.com/en/latest/upgrade/3.x.html#upgrade-from-3-0-to-3-1), 04.12.2016: minor release (age grade, heat index, new tools)
 * warning: requires at least php 5.6
* [v3.0](https://blog.runalyze.com/en/allgemein-en/runalyze-v3-0-en/), [update instructions](https://docs.runalyze.com/en/latest/upgrade/3.x.html#upgrade-from-2-6-to-3-0), 28.08.2016: major release (improved core, no amazing new features)
 * warning: requires at least php 5.5.9
* [v2.6](https://blog.runalyze.com/en/allgemein-en/runalyze-v2-6-en/), [update instructions](http://docs.runalyze.com/en/latest/upgrade/2.x.html#upgrade-from-2-5-to-2-6), 22.05.2016: minor release (race results, improved weather cache ...)
* [v2.5](http://blog.runalyze.com/en/allgemein-en/runalyze-v2-5-en/), [update instructions](http://docs.runalyze.com/en/latest/upgrade/2.x.html#upgrade-from-2-4-to-2-5), 12.04.2016: minor release (timezone support, Moving average, ...)
* [v2.4](http://blog.runalyze.com/en/allgemein-en/runalyze-v2-4-en/), [update instructions](http://docs.runalyze.com/en/latest/upgrade/2.x.html#upgrade-from-2-3-to-2-4), 25.01.2016: minor release (more weather data, recognition of duplicate activities, ...)
* [v2.3](http://blog.runalyze.com/en/allgemein-en/runalyze-v2-3-en/), [update instructions](http://docs.runalyze.com/en/latest/upgrade/2.x.html#upgrade-from-2-2-to-2-3), 11.12.2015: minor release (dataset refactoring, new running dynamics, ...)
 * warning: migration from v2.2 to v2.3 requires `refactor-geohash.php`
* [v2.2](http://blog.runalyze.com/en/allgemein-en/runalyze-v2-2-en/), [update instructions](http://docs.runalyze.com/en/latest/upgrade/2.x.html#upgrade-from-2-1-to-2-2), 28.10.2015: minor release (equipment for all sports, imperial units, ...)
 * If you're updating from 2.1.* directly use v2.2.1 for updating
 * warning: migration from v2.1 to v2.2 requires `refactor-equipment.php`
* [v2.1](http://blog.runalyze.com/en/allgemein-en/runalyze-v2-1-en/), 19.07.2015: minor release (running dynamics, new importers, recovery time ...)
* [v2.0](http://blog.runalyze.com/allgemein/runalyze-v2-0/), 28.02.2015: first mutlilingual major release
 * warning: migration from v1.5 to v2.0 requires `refactor-db.php` (see [v2.0alpha](http://blog.runalyze.com/allgemein/runalyze-v2-0alpha/), [v2.0beta](http://blog.runalyze.com/allgemein/runalyze-v2-0beta/))
* old versions (german only)
 * [v1.5](http://blog.runalyze.com/allgemein/runalyze-v1-5/), 01.01.2014: bugfixes, improved vdot formula
 * [v1.4](http://blog.runalyze.com/allgemein/runalyze-v1-4-fix-fuer-sicherheitsproblem/), 23.08.2013: bugfix for security issue
 * [v1.3](http://blog.runalyze.com/allgemein/runalyze-v1-3/), 29.07.2013: new importer, more data, vdot correction by elevation, ...
 * [v1.2](http://blog.runalyze.com/allgemein/runalyze-v1-2/), 13.11.2012: save plots, share activity list
 * [v1.1](http://blog.runalyze.com/allgemein/runalyze-v1-1/), 19.07.2012: first online version
 * [v1.0](http://blog.runalyze.com/allgemein/runalyze-v1-0/), 20.01.2012: first public version


inc/core/Dataset/Keys.php
<?php
/**
 * This file contains class::Keys
 * @package Runalyze
 */

namespace Runalyze\Dataset;

use Runalyze\Common\Enum\AbstractEnum;

/**
 * Enum for dataset keys and their internal ids
 *
 * @author Hannes Christiansen
 * @package Runalyze\Dataset
 */
final class Keys extends AbstractEnum
{
	/** @var int */
	const SPORT = 1;

	/** @var int */
	const TYPE = 2;

	/** @var int */
	const DAYTIME = 3;

	/** @var int */
	const SHARED_LINK = 4;

	/** @var int */
	const DISTANCE = 5;

	/** @var int */
	const DURATION = 6;

	/** @var int */
	const PACE = 7;

	/** @var int */
	const ELAPSED_TIME = 8;

	/** @var int */
	const ELEVATION = 9;

	/** @var int */
	const ENERGY = 10;

	/** @var int */
	const HEARTRATE_AVG = 11;

	/** @var int */
	const HEARTRATE_MAX = 12;

	/** @var int */
	const VO2MAX_ICON = 13;

	/** @var int */
	const VO2MAX_VALUE = 14;

	/** @var int */
	const FIT_VO2MAX_ESTIMATE = 15;

	/** @var int */
	const FIT_RECOVERY_TIME = 16;

	/** @var int */
	const FIT_HRV_ANALYSIS = 17;

	/** @var int */
	const TRIMP = 19;

	/** @var int */
	const CADENCE = 20;

	/** @var int */
	const POWER = 21;

	/** @var int */
	const SWOLF = 22;

	/** @var int */
	const STRIDE_LENGTH = 23;

	/** @var int */
	const GROUNDCONTACT = 24;

	/** @var int */
	const VERTICAL_OSCILLATION = 25;

	/** @var int */
	const TEMPERATURE = 26;

	/** @var int */
	const WEATHER = 27;

	/** @var int */
	const ROUTE = 28;

	/** @var int */
	const SPLITS = 29;

	/** @var int */
	const TITLE = 30;

	/** @var int */
	const TRAININGPARTNER = 31;

	/** @var int */
	const TOTAL_STROKES = 32;

	/** @var int */
	const VERTICAL_RATIO = 33;

	/** @var int */
	const GROUNDCONTACT_BALANCE = 34;

	/** @var int */
	const TAGS = 35;

	/** @var int */
	const COMPLETE_EQUIPMENT = 36;

	/** @var int */
	const MAIN_EQUIPMENT = 37;

	/** @var int */
	const WIND = 38;

	/** @var int */
	const HUMIDITY = 39;

	/** @var int */
	const AIR_PRESSURE = 40;

	/** @var int */
	const WIND_CHILL = 41;

	/** @var int */
	const SETTING = 42;

	/** @var int */
	const FIT_TRAINING_EFFECT = 43;

	/** @var int */
	const RPE = 44;

	/** @var int */
	const GRADIENT = 45;

	/** @var int */
	const FIT_PERFORMANCE_CONDITION = 46;

	/** @var int */
	const RACE_RESULT = 47;

	/** @var int */
	const HEAT_INDEX = 48;

    /** @var int */
    const FIT_PERFORMANCE_CONDITION_START = 49;

    /** @var int */
    const FIT_PERFORMANCE_CONDITION_END = 50;

    /** @var int */
    const CLIMB_SCORE = 51;

    /** @var int */
    const PERCENTAGE_HILLY = 52;

    /** @var int */
    const TOOLS = 53;

    /** @var int */
    const FLIGHT_TIME = 54;

    /** @var int */
    const FLIGHT_RATIO = 55;

    /** @var int */
    const AVG_IMPACT_GS = 56;

    /** @var int */
    const AVG_BRAKING_GS = 57;

    /** @var int */
    const AVG_FOOTSTRIKE_TYPE = 58;

    /** @var int */
    const AVG_PRONATION_EXCURSION = 59;

	/** @var int */
	const FIT_ANAEROBIC_TRAINING_EFFECT = 60;

    /**
	 * @var array|null
	 */
	private static $ClassNames = null;

	/**
	 * @var array instances of key objects
	 */
	private static $Instances = array();

	/**
	 * @var bool
	 */
	private static $KeepInstances = false;

	/**
	 * @param bool $flag
	 */
	public static function keepInstances($flag = true)
	{
		self::$KeepInstances = $flag;
	}

	/**
	 * Get key
	 * @param int $keyid int from internal enum
	 * @return \Runalyze\Dataset\Keys\AbstractKey
	 * @throws \InvalidArgumentException
	 */
	public static function get($keyid)
	{
		if (null === self::$ClassNames) {
			self::generateClassNamesArray();
		}

		if (!isset(self::$ClassNames[$keyid])) {
			throw new \InvalidArgumentException('Invalid keyid "'.$keyid.'".');
		}

		$className = 'Runalyze\\Dataset\\Keys\\'.self::$ClassNames[$keyid];

		if (self::$KeepInstances) {
			if (!isset(self::$Instances[$keyid])) {
				self::$Instances[$keyid] = new $className;
			}

			return self::$Instances[$keyid];
		}

		return new $className;
	}

	/**
	 * Generate static array with class names
	 */
	private static function generateClassNamesArray()
	{
		self::$ClassNames = array(
			self::SPORT => 'Sport',
			self::TYPE => 'Type',
			self::DAYTIME => 'Daytime',
			self::SHARED_LINK => 'SharedLink',
			self::DISTANCE => 'Distance',
			self::DURATION => 'Duration',
			self::PACE => 'Pace',
			self::ELAPSED_TIME => 'ElapsedTime',
			self::ELEVATION => 'Elevation',
			self::ENERGY => 'Energy',
			self::HEARTRATE_AVG => 'HeartrateAverage',
			self::HEARTRATE_MAX => 'HeartrateMaximum',
			self::VO2MAX_ICON => 'VO2maxIcon',
			self::VO2MAX_VALUE => 'VO2maxValue',
			self::FIT_VO2MAX_ESTIMATE => 'FitVO2maxEstimate',
			self::FIT_RECOVERY_TIME => 'FitRecoveryTime',
			self::FIT_HRV_ANALYSIS => 'FitHrvAnalysis',
			self::TRIMP => 'Trimp',
			self::CADENCE => 'Cadence',
			self::POWER => 'Power',
			self::SWOLF => 'Swolf',
			self::STRIDE_LENGTH => 'StrideLength',
			self::GROUNDCONTACT => 'Groundcontact',
			self::VERTICAL_OSCILLATION => 'VerticalOscillation',
			self::TEMPERATURE => 'Temperature',
			self::WEATHER => 'Weather',
			self::ROUTE => 'Route',
			self::SPLITS => 'Splits',
			self::TITLE => 'Title',
			self::TRAININGPARTNER => 'TrainingPartner',
			self::TOTAL_STROKES => 'TotalStrokes',
			self::VERTICAL_RATIO => 'VerticalRatio',
			self::GROUNDCONTACT_BALANCE => 'GroundcontactBalance',
			self::TAGS => 'Tags',
			self::COMPLETE_EQUIPMENT => 'CompleteEquipment',
			self::MAIN_EQUIPMENT => 'MainEquipment',
			self::WIND => 'Wind',
			self::HUMIDITY => 'Humidity',
			self::AIR_PRESSURE => 'AirPressure',
			self::WIND_CHILL => 'WindChill',
			self::SETTING => 'Setting',
			self::FIT_TRAINING_EFFECT => 'FitTrainingEffect',
			self::RPE => 'RPE',
			self::GRADIENT => 'Gradient',
			self::FIT_PERFORMANCE_CONDITION => 'FitPerformanceCondition',
			self::RACE_RESULT => 'RaceResult',
			self::HEAT_INDEX => 'HeatIndex',
            self::FIT_PERFORMANCE_CONDITION_START => 'FitPerformanceConditionStart',
            self::FIT_PERFORMANCE_CONDITION_END => 'FitPerformanceConditionEnd',
            self::CLIMB_SCORE => 'ClimbScore',
            self::PERCENTAGE_HILLY => 'PercentageHilly',
            self::TOOLS => 'Tools',
            self::FLIGHT_TIME => 'FlightTime',
            self::FLIGHT_RATIO => 'FlightRatio',
            self::AVG_IMPACT_GS => 'AverageImpactGs',
            self::AVG_BRAKING_GS => 'AverageBrakingGs',
            self::AVG_FOOTSTRIKE_TYPE => 'AverageFootstrikeType',
            self::AVG_PRONATION_EXCURSION => 'AveragePronationExcursion',
            self::FIT_ANAEROBIC_TRAINING_EFFECT => 'FitAnaerobicTrainingEffect'
        );
	}
}
